### DISCUSSION TITLE
Forcing off time at each end of the cycle

### DISCUSSION THREAD

**Discussion by ms5** [jmcollin78/versatile_thermostat#602] [Ideas]
I am looking for a solution to force the thermostat OFF at the end of each cycle for some time.  I currently have Legrand analog wall thermostats installed and they work with a PWM where after each cycle the relais stays off for some time. On the Legrand thermostats this is not really configurable, but I was able to measure what happens and the intervals by connecting a shelly to the thermostat and watch it for some time. 

I thought about how to achieve this and a very easy solution that I came up with, is to multiply the calculated ON seconds with a configurable multiplicator to limit the max. ON time to a certain amount per cycle. Here an example: The cycle is set to 10min, the multiplicator is set to 0.75, if VT calculates an on time of 100% = 600seconds, this is then multiplied by 0.75 so we get 450seconds ON time and 150seconds OFF time. 

![Screenshot 2024-11-04 at 15 40 20](https://github.com/user-attachments/assets/bc3e1210-46a9-4aef-8c21-d5cd6bf9e974)

I have attached a little graph that show in red the ON time and in blue the min. forced OFF time. The example uses the above parameters. 

I tested this solution over the last weekend and it seems to work very well. For my example (to replace the Legrand Thermostats) I used a cycle of 10min, a min activation delay of 3min and a multiplicator of 0.6. This gives me an ON time for a heating cycle between 3 and 6 minutes and at least 4min OFF at the end of each cycle. The times still vary depending on the calculation, but at least 4min OFF at the end of each cycle are guaranteed. 

I have made a POC of this that can be found here: https://github.com/ms5/versatile_thermostat/tree/output_time_modifier

The configuration can be made globally in `configuration.yaml`
```
versatile_thermostat:
    output_time_modifier:
        multiplicator: 0.75
```

I would like to get some inputs and hopefully get this integrated.

**Comment by jmcollin78** [comment]
Hello @ms5 ,

By doing this you are cheating with the TPI algorithm and then you will not be able to reach your target temperature.
This is not something, I recommend.

My advice: change the faulty thermostat. I consider asking to change an algorithm because of a faulty equipment is a very bad way to fix an issue.

So, you understand, I will not integrate this. Of course you can keep your fork if you want.

**Reply by ms5** [reply]
Hi @jmcollin78 

why do you think the equipment is faulty? It is the expected behavior of it. The heating system is slow, too much ON time will waste more energy than a shorter on off cycle. The off part is important to make it energy efficient. There is also no other option to force time off after a ON period, not is there an option to black activation in the next cycle when the switch was on for some time. If the temperature is a couple degrees away from the set point, VT will keep the switch just on for a couple of cycles with the result that energy is wasted.

**Reply by jmcollin78** [reply]
> where after each cycle the relais stays off for some time.

I suppose this was a faulty behavior, but maybe I interpret badly your words.

I wonder if a VTherm `over_switch` is a good solution for your case. Do you have another `climate` entity that can control your thermostat instead of a switch ?

**Reply by ms5** [reply]
Its a complex system and I can elaborate a bit how it works, but (as I found out myself) it's quiet a deep rabbit hole. 

The system is a "low temperature underfloor waterheating". Low temperature means the spread between room temperature and heating water is also quite low. I sometime measure less than 8°C difference.  In my case this one system covers 4 houses and about 50 apartments, so I do not have any control over the heatpump nor the temperature, etc. The system is controlled with a valve (in a central box) for each room. The used valves do only know the positions ON or OFF. It also takes about 3min from powering the valve to ON till water flows  and the same amount of time is used to turn it off again. 

According my the HVAC contractor they do not limit flow to each room, but let the thermostat algorithm take care of it. As he explained it to me it works like this: Thermostat switches relay to ON -> valve gets power -> 3min later water starts to flow with about 2.5 liters per minute -> after some time (depending on the thermostats algorithm between 3 and 8minutes)  relay switches off -> valve loses power -> 3min later water flow stops. Now here comes the important part, as the HAVC contractor explained it to me: now we have warm water sitting in the underfloor pipes below the room. Since the temperature spread is low, it takes time for the floor to absorb the heat provided. Therefore the off cycle. But as the contractor explained: this is part of the heating phase and during this "pause" heat is absorbed and the room gets warmer. Then after the pause and if the set temperature has not been reached, the cycle continues, maybe with slightly different ON/OFF times. This is also the reason they installed these Legrand thermostats, since they are algorithm  based and not just ON/OFF hysteresis switches. 
Then if you skip the OFF phase, the still warm water gets pushed out the pipe too early and the remaining heating energy is basically lost. 

I acquired this information from the contractor recently, as I am in the process of figuring out, why I used around 60% more energy last winter (with the normal VT) compared to the year before when I still used the Legrand thermostats. 

Legrand unfortunately does not explain what they exactly do. In their datasheet they have this phrase:
`The regulation is ensured by an algorithm (commutation cycle) thus ensuring a minimal temperature variation relative to the ambient temperature`. I found this not very helpful and I wanted to know more about it, therefore I connected a shelly device to the output relay of such a thermostat to see what ON/OFF cycles it has and how often it does things. What I figured is that it's basically a PWM signal of around 10min which never reached 100%. It is capped at around 60% and also has a minimal activation delay of around 3min. This is where my parameters come from. I think the Legrand product is actually solid and works very well, unfortunately it is all analog and can only be controlled with a knob on the device itself. This is why I am here. 

So back to your question: I do not see any benefit of using a different climate entity, since the only thing I actually can control is the ON/OFF situation of the valve. The valve cannot open 20% or 50%. It only as 0% or 100%.

**Reply by jmcollin78** [reply]
If you use the native climate entity which implement the specificities and put a VTherm above it, you will have the specificities and also the features of VTherm without to implement the specificities into VTherm. This is best solution when you have some specificities.

**Reply by ms5** [reply]
How would that help me?  I need the VT to calculate the ON time percent, then I need to scale this value to fit my HVAC system (since the multiplication). If I put a VT behind another Thermostat I will lose this. I have no clue why you recommend this, maybe you can elaborate. 

I have also checked how other projects approached this and for example smart pid thermostat has an output clamp setting, which limits the output to a max value (I till think the scaling is better, but this is certainly an approach that addresses the issue). 
I also belief I am not the only one with such a system. True this setting is not needed for most, but there are use cases. I also belief some just give up, because they cannot find a system that actually fits the installed HVAC.

**Reply by jmcollin78** [reply]
If you have a climate that calculate for you the onPercent and do the correct specificities of your device, then you just put a VTherm over it, and voilà.

> I have also checked how other projects approached this and for example smart pid thermostat has an output clamp setting, which limits the output to a max value

This is an option. Having a max value is something easy to do (but in expert mode to avoid over complexify the configuration)

**Reply by ms5** [reply]
> If you have a climate that calculate for you the onPercent and do the correct specificities of your device, then you just put a VTherm over it, and voilà.

If I have to write a complete new thermostat, what is the point of using VT. I still don't see the benefit, since everything I use from VT would have to be implemented in the new entity too. 

I am personally also more a fan of bundling efforts into a project, than making 100small ones that will die soon anyways. Versatile thermostat is therefore a great project and you already put so much work into it.


> This is an option. Having a max value is something easy to do (but in expert mode to avoid over complexify the configuration)

My proposal has this settings in the config file. I think, chances are small someone could enable it accidentally and defaults are not changed. I can certainly rename it to something "expert". The already existing "expert" mode `auto_regulation_expert ` is IMHO more related to regulation so I will not put it under there. I think a new section will be clearer. What would you thing about this?

```
versatile_thermostat:
    expert_output_on_time_modifiers:
        on_time_scaling: 0.75 # [float, default to 1]
        on_time_clamping_high_percentage: 0.75 #[float, default to 1]
```

I have also made a graph to visualize the outcome of the two different modifiers. I personally belief that scaling will deliver better results, since the scale of the calculated value will be kept. Although I have to admit I have not tested clamping yet. I belief there are use cases for both. 


![Screenshot 2024-11-07 at 08 06 16](https://github.com/user-attachments/assets/af041a8a-09b5-45a8-bbce-63c4873add6c)

**Reply by jmcollin78** [reply]
> If I have to write a complete new thermostat, what is the point of using VT

That is not my point. I suppose you already have an climate entity given by your thermostat vendor to handle the specificities. If yes,  you just have to put a VTherm over it (over_cliamte).
Else, my words make no sense and you can safely ignore it.

> I think a new section will be clearer. What would you thing about this?

Yes fine. If have another proposal: have a max value for the on_percent. The on_percent is calculated by the TPI algorithm and give the percent of time the heater should be on. If we clamp it to a max value < 1, you will always have an off perid equal to (1 - on_percent) x cycle_min minute. And so we just have one value to set with something like that (no need of subsection):

```
versatile_thermostat:
    max_on_percent: 0.8
```

But once more, you will generate disturbances in the heating regulation and could never reach the target temperature with such a parameter.

**Reply by ms5** [reply]
That is perfect and will work. I will implement it and create a pull request. Probably somewhere along this week.

**Reply by jmcollin78** [reply]
ok fine. Please start from the last beta else you will have a big merge. I will release it soon.
